cluster.go

```diff
@@ -27,192 +26,152 @@ import (
+		}
+		// step: set the default protocol schema
+		if defaultProto == "" {
+			if u.Scheme != "http" && u.Scheme != "https" {
+				return nil, ErrInvalidEndpoint
+			}
+			defaultProto = u.Scheme
+		}
+		// step: does the url have a protocol schema? if not, use the default
+		if u.Scheme == "" {
+			u, _ = url.Parse(fmt.Sprintf("%s://%s", defaultProto, u.String()))
+		}
 
-func newCluster(client *http.Client, marathonURL string) (Cluster, error) {
-	// step: parse the marathon url
-	marathon, err := url.Parse(marathonURL)
-	if err != nil {
-		return nil, ErrInvalidEndpoint
-	}
+		// step: check for empty hosts
+		if u.Host == "" || u.Scheme == "" {
+			return nil, ErrInvalidEndpoint
+		}
 
-	// step: check the protocol
-	if marathon.Scheme != "http" && marathon.Scheme != "https" {
-		return nil, ErrInvalidEndpoint
+		// step: create a new node for this endpoint
+		members = append(members, &member{endpoint: u.String()})
 	}
 
-	cluster := &marathonCluster{
-		client:   client,
-		protocol: marathon.Scheme,
-		url:      marathonURL,
-	}
+	return &cluster{
+		client:  client,
+		members: members,
+	}, nil
+}
 
-	/* step: create a link list of the hosts */
-	var previous *marathonNode
-	for index, host := range strings.SplitN(marathon.Host+marathon.Path, ",", -1) {
-		if len(host) == 0 {
-			return nil, ErrInvalidEndpoint
+// retrieve the current member, i.e. the current endpoint in use
+func (c *cluster) getMember() (string, error) {
+	c.RLock()
+	defer c.RUnlock()
+	for _, n := range c.members {
+		if n.status == memberStatusUp {
+			return n.endpoint, nil
 		}
+	}
 
-		// step: create a new cluster member
-		node := new(marathonNode)
-		node.hostname = host
-		cluster.size++
-		// step: if the first member
-		if index == 0 {
-			cluster.members = node
-			cluster.active = node
-			previous = node
-		} else {
-			previous.next = node
-			previous = node
+	return "", ErrMarathonDown
+}
+
+// markDown marks down the current endpoint
+func (c *cluster) markDown(endpoint string) {
+	c.Lock()
+	defer c.Unlock()
+	for _, n := range c.members {
+		// step: check if this is the node and it's marked as up - The double  checking on the
+		// nodes status ensures the multiple calls don't create multiple checks
+		if n.status == memberStatusUp && n.endpoint == endpoint {
+			n.status = memberStatusDown
+			go c.healthCheckNode(n)
+			break
 		}
 	}
-	// step: close the link list
-	previous.next = cluster.active
-
-	return cluster, nil
 }
 
-func (r *marathonCluster) URL() string {
-	return r.url
+// healthCheckNode performs a health check on the node and when active updates the status
+func (c *cluster) healthCheckNode(node *member) {
+	// step: wait for the node to become active ... we are assuming a /ping is enough here
+	for {
+		res, err := c.client.Get(fmt.Sprintf("%s/ping", node.endpoint))
+		if err == nil && res.StatusCode == 200 {
+			break
+		}
+		<-time.After(time.Duration(5 * time.Second))
+	}
+	// step: mark the node as active again
+	c.Lock()
```

Comment:
We could also pull this section right into the 200 block (i.e., line 130) and return from the function afterwards.